doc/fluid/design/dist_train/distributed_lookup_table_design.md

```diff
@@ -44,111 +44,39 @@ $$W = f(W, W')$$
-1. For some cases, the range of input `Ids` is very wide and unpredictable, so the sparse
-table would be able to fill a new value for the id that didn't appear before with
-zero, uniform random or Gaussian distribution.
-
-For each Trainer's training process:
-1. In the forward pass, we use `pre-fetch` op to pre-fetch parameter blocks according to the
-input `Ids` from PServers instead of the local `lookup_table` op, and then merge the blocks
-into a parameter `W`.
-1. Compute `GRAD@W'` in the backward pass using the pre-fetched `W` and send it to PServer to
-execute the optimize pass.
-
-## Conclusion
-
-Let us do the "storage service does not optimize" solution first, as a
-baseline at least, because it is easier to use a well-optimized
-distributed storage service like memcached.  We can do the "storage
-service does optimize" solution later or at the same time, which, if
-implemented carefully, should have better performance than the former.
+## Distributed Lookup Table
+### Problem 1: The lookup table may be very large.
+
+ In the condition like the search engine and recommendation system, the number of feature ID may be very large, see 1000000000, then for a lookup table of size 8, the total size of the table is:
+
+ ```
+ 100000000000 * 8 * 4.0 = 2980.23 GB
+ ```
+
+### Solution: Distributed storage
+
+1. Paddle use SelectedRows as the storage format for the lookup table, the lookup table parameter will be split to multi-machine according to the hash of the feature ID, and data will also be split and send to the same machine to prefetch the parameter.
+
+1. For common parameters, the trainer will get the whole parameter for training, but for the big lookup table, the trainer can not store the whole parameter, but the input data feature is very sparse, so every time we only need a few parameters for training, so we use `prefetch_op` to only prefetch the parameter needed to trainer.
+
+### Problem 2. The Id in the lookup table is not sure before training.
+
+ The feature Id is calculated by the hash function because the feature data source is so large, we can not get all the id before training. So we can not initialize the table before training.
+
+### Solution: Id auto growth
+
+At the beginning of training, paddle only malloc the memory for the lookup table at parameter side, the id and the data will not be initialized. During training, when a parameter server received an Id, if it is already in the lookup table, it will return the existing parameter, if the id does not exist, paddle will add it into the lookup table and initialize the value for it.
+
+
+## Architecture
+The whole architecture of the distribute lookup table is as below:
+
+### Training steps:
+1. Read a batch of data, the data is feature ids.
+1. The input ids will be split by `split_ids_op` with the same hash function of the lookup table.
+1. The `prefetch_op` use the split result to prefetch parameters back from the lookup table.
```

Comment:
Maybe we need a fault-tolerance design especially for distributed lookup table?

Reply:
Good point!  Currently, we can only support a fix sized lookup table, we need to estimate the size now. One solution can be that the table supports auto memory growth.

Reply:
We can start the design at next stage, after we can train a industrial model.